alacritty/src/window_context.rs

```diff
@@ -0,0 +1,345 @@
+        // The display manages a window and can draw the terminal.
+        let display = Display::new(config, window_event_loop)?;
+
+        info!(
+            "PTY dimensions: {:?} x {:?}",
+            display.size_info.screen_lines(),
+            display.size_info.columns()
+        );
+
+        let event_proxy = EventProxy::new(proxy, display.window.id());
+
+        // Create the terminal.
+        //
+        // This object contains all of the state about what's being displayed. It's
+        // wrapped in a clonable mutex since both the I/O loop and display need to
+        // access it.
+        let terminal = Term::new(config, display.size_info, event_proxy.clone());
+        let terminal = Arc::new(FairMutex::new(terminal));
+
+        // Create the PTY.
+        //
+        // The PTY forks a process to run the shell on the slave side of the
+        // pseudoterminal. A file descriptor for the master side is retained for
+        // reading/writing to the shell.
+        let pty = tty::new(config, &display.size_info, display.window.x11_window_id());
+
+        // Create the pseudoterminal I/O loop.
+        //
+        // PTY I/O is ran on another thread as to not occupy cycles used by the
+        // renderer and input processing. Note that access to the terminal state is
+        // synchronized since the I/O loop updates the state, and the display
+        // consumes it periodically.
+        let event_loop = PtyEventLoop::new(
+            Arc::clone(&terminal),
+            event_proxy.clone(),
+            pty,
+            config.hold,
+            config.ui_config.debug.ref_test,
+        );
+
+        // The event loop channel allows write requests from the event processor
+        // to be sent to the pty loop and ultimately written to the pty.
+        let loop_tx = event_loop.channel();
+
+        // Kick off the I/O thread.
+        let _io_thread = event_loop.spawn();
+
+        // Start cursor blinking, in case `Focused` isn't sent on startup.
+        if config.cursor.style().blinking {
+            event_proxy.send_event(TerminalEvent::CursorBlinkingChange.into());
+        }
+
+        // Create context for the Alacritty window.
+        Ok(WindowContext {
+            font_size: config.ui_config.font.size(),
+            notifier: Notifier(loop_tx),
+            terminal,
+            display,
+            suppress_chars: Default::default(),
+            message_buffer: Default::default(),
+            received_count: Default::default(),
+            search_state: Default::default(),
+            event_queue: Default::default(),
+            modifiers: Default::default(),
+            mouse: Default::default(),
+            dirty: Default::default(),
+        })
+    }
+
+    /// Update the terminal window to the latest config.
+    pub fn update_config(&mut self, old_config: &Config, config: &Config) {
+        self.display.update_config(config);
+        self.terminal.lock().update_config(config);
```

Comment:
The terminal lock here could block for a bit, but in theory a config update should hopefully be rare enough to make this a non-issue.